### ISSUE TITLE
Site 068 Portland - Security Deposit management - Deposit disappeared

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1764]
In GitLab by @kdjstudios on Nov 11, 2019, 09:47

**Submitted by:** <jeffrey.casey@answernet.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2019-11-08-24865

**Server:** Internal

**Client/Site:** Portland

**Account:** AA2613

**Issue:** 

Account number 068-AA2613.  Working up a final invoice that we applied the deposit to the invoice balance retaining the remainder.  Added an additional line item to the invoice and the deposit balance disappeared.  Canceled the invoice keeping the deposit did not bring it back.  The deposit on the account is $115.00.  I have to get this invoice out therefore I am managing the deposit amount with stage fees as we have in the past.  Would like to understand what we possibly did incorrectly for this to happen?

`SD TEST CASES:`: [URL](https://docs.google.com/spreadsheets/d/136cTbTgOhSdRoilK_j-6qbm6F2MQ9T-w2lWbeOiSZa4/edit#gid=0)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 11, 2019, 09:47

I wrote:

> Please do NOT use staged fees to track deposits. That will not work. Please let me know if you have time today for a call to review how to use the deposit system in SA Billing.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 12, 2019, 14:41

@smasih @amishra @jredfield 

After a phone call with Jeffrey, he was able to explain what he attempted to do, and what SA Billing actually did. 

Attempted to apply to account balance and and refund remain to invoice. There was 115 in the deposit to start with, which 58.90 should have went to zero out the account, and then 56.10 should be refunded manually by operations.

However, we are getting an error when trying to access the Security Deposit Fees section.

![image](/uploads/1553d86b1d2f0e8fe88585c0bc5eabd0/image.png)

In the end he went in and manually adjusted the invoice with a staged fee, which we need to change to be the actual SD fee. So the reports show correctly.  

Steps Jeffrey stated he performed: (I was also able to reproduce this issue too.)
1. Go to draft invoice
2. Choose the manage deposit
3. Then choose additional fee and add it. Then remove the additional fee.
4. This results in the line items for SD zeroing out.
5. Then deleted the invoice to attempt to undo, which it did not. 

![image](/uploads/254e629191aa443d36a86ee40abf9fe8/image.png)

![image](/uploads/02317b3197b3c520a0fd816569cad9ab/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 12, 2019, 14:42

I believe we will need to perform the following:

1. Manually update the specified account so it shows correctly. We will need to have it show that 58.90 of the SD was applied to the account balance and then also show that 56.10 is being returned to the client. 
2. Find what is causing this and estimate a solution to correct it. My presumption is it has something to do with the functionality when deleted a fee on the draft invoice.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 22, 2019, 04:16

@kdjstudios @jredfield 

Thank you for the update.

> However, we are getting an error when trying to access the Security Deposit Fees section.

* We looked into this and found that the code was missing/blank of the Security Deposit Stage Fee.
* We tried to find the root cause of this why the code was nil/blank but unable to found the root cause yet.

* **Quick update for Security Deposit Fees section page:** We need to update the code(like 9902,9903..etc..) from the security deposit Stage Fee. or just add one condition if code is present else show blank.

**What step do we need to do for this account? i.e what should have the account & security deposit balance**

`NOTE: TOMORROW WE ARE GOING TO REMOVE OLD SECURITY DEPOSIT FROM SAB.`

> Find what is causing this and estimate a solution to correct it. My presumption is it has something to do with the functionality when deleted a fee on the draft invoice.

yes, you are right when SD is present and add/remove the line item from the draft invoice SD 
 the calculation is not working properly. I'll discuss with @smasih  and let you know what's the solution.



Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 27, 2019, 10:46

@amishra @smasih @jredfield 

Why do we not have an update on this ANP0.5 ticket in over a month?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 27, 2019, 13:28

Jeffrey wrote in a new ticket:

> Not sure what is happening.  This account is scheduled to be reactivated.
> The account had a $80.80 deposit showing.  We now have a Deposit Decrease
> line item on our draft invoice at $0.00.  The account now shows the clinet
> with $0.00 deposit.  Unable to find any pending deposit transaction.  The
> deposit does not appear to have been applied to any account balances.
> 
>  
> 
> We put in a ticket that back in November very closely related to this
> behavior ( 08 Nov 2019 - Site 068 Portland - Security Deposit management -
> Deposit disappeared ) on an inactive account.  This request is not more
> urgent due to it being on an active account that will be going live once
> again in the system.  
> 
>  
> 
> Where did this clients deposit go?  I do not believe that we requested any
> movement of the deposit? What is happening? 

I wrote:

> Please be sure in the future to respond to your original ticket. Opening new tickets only causes delays and confusion. 
> 
> In regards to providing an update, the support team is still researching this and there is no ETA yet for a fix. We will keep you updated on any progress made.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 30, 2019, 04:51

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Dec 30, 2019, 06:22

@kdjstudios @jredfield 

We have looked into this but did not find the root cause yet. We need to dig deeper into this. 

But in the meantime, we can fix this account manually. For which we need the SD code which should be applied to the Security Deposit Stage Fee and the amount of security Deposit balance of the account.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 30, 2019, 09:57

Jeffrey wrote:

> My apologies for leaving out some critical information.  This is not a follow up of our original ticket.  This ticket is reporting three more instances of this behavior:
> 
> Account 068-AA2642 - Gordon truck center - $270.??
> 
> Account 068-GM1010 - Linage apartments - $80.80
> 
> Account 068-B06351 - Jefferson COA - $34.95

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 30, 2019, 09:58

@ranand 

Please make sure all of the above accounts have the SD placed fully back into the SD balance and there are no pending SD fees. I will then review with Jeffrey and attempt to get SAB to actually work.

I wrote:

> For all of the accounts mentioned, we will have the support team remove any pending SD fees and ensure that the SD balance is brought back to what it was. At that point you and I will need to jump on a zoom/call to review the process for you to take to manage the deposits. 

Jeffrey wrote:

> That sounds like a plan.  Let us know when you would like to get together and go over the process.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 1, 2020, 23:12

@kdjstudios @jredfield 

Thank you for the update.

* Account 068-AA2642 - Gordon truck center(**Active**) - `$270`
 
* Account 068-GM1010 - Linage apartments(**Inactive**) - `$80.80`
 
* Account 068-B06351 - Jefferson COA(**Inactive**) - `$34.95`

* Account 068-AA2613 CODA'S Plus(**Inactive**) - **What amount we need to update?**    
  * **For Security Deposit Fees section page:** there is missing code under the `Finalized Security Deposit Fees` on Security Deposit DFee Section. We need to update the code(like 9902,9903..etc..) for the security deposit Stage Fee in order to open the page or just add one condition if code is present else show blank.

* **Please confirm for all account SD amount so that we will do this directly on the production server.**

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 14, 2020, 06:22

@kdjstudios @jredfield 

I am still waiting for your response on my last comment.
Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 14, 2020, 08:49

@amishra @jredfield 

Thank you for bringing this back to our attention. I did not see your comment above. 

I wrote:

> The support team was able to review the logs and believes the following values were the Security Deposit amount for each account. Please review and confirm these amounts. Once we have your confirmation we will have the support team make the corrections to the Security deposit balances.  
> 
> * Account 068-AA2642 - Gordon truck center(Active) - $270
> * Account 068-GM1010 - Linage apartments(Inactive) - $80.80
> * Account 068-B06351 - Jefferson COA(Inactive) - $34.95
> * Account 068-AA2613 CODA'S Plus(Inactive) - $115.00


**For Security Deposit Fees section page:** there is missing code under the `Finalized Security Deposit Fees` on Security Deposit DFee Section. We need to update the code(like 9902,9903..etc..) for the security deposit Stage Fee in order to open the page or just add one condition if code is present else show blank.

> May you please explain how this happened? Also what do you mean add one condition if code is present? You may need to discus this with @jredfield during the morning meetings for clarifications.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Jan 22, 2020, 01:20

@kdjstudios @jredfield 

> May you please explain how this happened?

We did not find the root cause yet.

> Also what do you mean add one condition if code is present?

The issue here is we have a **security deposit Stage Fee** on the account for which the **SD Code**(ie. 9903, 9902,9998,9997) is missing. Because of this, SAB is unable to open **Security Deposit Fees section page** for the account, as it always expects the **SD Code** for the **security deposit Stage Fee**. So, @amishra proposed that we will add the check for **SD Code** presence on the **Security Deposit Fees section page**.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 23, 2020, 12:19

@ranand 

Thank you, yes please provide an estimate to add that condition. If less then 2 hours, it is pre-approved.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 29, 2020, 03:48

@kdjstudios @jredfield 

Thank you for the update.

We have fixed the code & now we are able to see the Security Deposit Stage Fee page.   

We have setup a branch on the merge2 server. Please test it so that we will release this in our upcoming release.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 1, 2020, 01:26

@kdjstudios @jredfield 

We have successfully released on both of the production servers.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 4, 2020, 13:09

I wrote:

> We are just following up again as we have not received a response from our original email sent Jan 14th. Please review and confirm the amounts so we may begin to setup a meeting to review how to process the accounts.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 4, 2020, 15:08

Jeffrey wrote:

> Not sure how we missed your correspondence on the 14th.  Our apologies. 
> 
> Portland agrees with the amounts listed:
> * Account 068-AA2642 - Gordon truck center(Active) - $270 
> * Account 068-GM1010 - Linage apartments(Inactive) - $80.80 
> * Account 068-B06351 - Jefferson COA(Inactive) - $34.95 
> * Account 068-AA2613 CODA'S Plus(Inactive) - $115.00
> 
> We have had one additional anomaly:
> *  Account 068- AA6155 - C & R Real Estate Services - $315.00
> 
> Leah and I are looking forward to getting together to discuss the process.  When are we planning on getting together on this matter?
> 
> Appreciate your time on this matter.

I wrote:

> We have informed the support team about the additional account. I am expecting to have that account's SD reverted back by the end of the week. So lets look at next week for scheduling a meeting to review how to use the SD wizard in SAB. I have availability for either Wed. or Thurs. between 11am to 2pm MT. I would expect we should not need more then a one hour time block.  


![image](/uploads/4bfd5b964d72a8ca30aca50c922f0cd7/image.png)

@amishra @smasih 

Please correct this account 068- AA6155 too.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 4, 2020, 15:14

@amishra @smasih @ranand 

I just also want to confirm that we have still not fixed the issue outlined in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1633#note_51154; where basically when a user adds and/or removes an additional fee to the draft invoice it breaks the SD fees that are on that draft invoice, correct?

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Feb 4, 2020, 20:34

@kdjstudios @jredfield 

Yes, We still need to look into it.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 5, 2020, 08:22

removed milestone

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 5, 2020, 08:24

@ranand @jredfield 

Thank you for the update. Placing this back into development as it will need to be fixed. Keeping as important too. Please provide an estimate, but proceed with developing the fix.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 6, 2020, 23:51

@kdjstudios @jredfield 

Thank you for the update.

We have updated the SD balance for all accounts mentioned above in [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1633#note_55447) on the dev server first.

Please confirm with the client that everything showing as expected. so that we will do this same as on the production server.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 7, 2020, 09:47

@amishra @ranand @jredfield 

I am confused with your response, there was only one account to update? The other 4 had already been done previously in this ticket. 

Where is the estimate?

Please proceed with doing this on production.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 10, 2020, 24:15

@kdjstudios @jredfield 

Thank you for the update.

> I am confused with your response, there was only one account to update? The other 4 had already been done previously in this ticket.
 
* Yes, you are right, I just thought, we will do all accounts at the same time on the production server once the last one account confirms.


> Where is the estimate?

It will take around 40 min



> Please proceed with doing this on production.

We have fixed all accounts on the production server.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 10, 2020, 10:19

@amishra 

Thank you. We are still needing to fix the actual issue as mentioned in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1633#note_51154. Marking back to development.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 13, 2020, 08:44

This was reviewed today and @smasih believes they understand how to reproduce the issue, but will need further researching to determine what in the code is causing the issue. Also they expect once they do find the issue that testing the fix may take a bit of time as they will need to go through all the SD test cases to confirm no other unexpected behavior.

Please research and post and estimate.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 17, 2020, 05:06

@kdjstudios @jredfield 

Thank you for the update.

We looked into this and our findings are described below:

 
* The issue comes only when we choose a refund option from the draft invoice page then the security Balance becomes zero and when we add a staged fee from the draft invoice & delete them they put it the zero on the SD line item because the SD balance was zero.
* below is the short video of the actual scenarios of the issue.
  * [Short Video](https://www.loom.com/share/0bb8bb67c088435fa60819c97ea0239c)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 17, 2020, 11:48

@amishra @jredfield @smasih 

Thank you, however, I believe we already knew what the issue was, as it was pointed out in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1633#note_51154. We need to know what is causing this (What in the code is wrong) and what needs to be adjusted to fix it. Please provide an estimate.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 19, 2020, 05:44

@kdjstudios @jredfield 

We have figured the problem. We are updating the SD balance and the previous SD balance every time an additional fee is added/deleted/updated. 

You can check the file "https://github.com/ANProgrammers/sabilling_merge/blob/master/app/models/account.rb" at Line no. 469, it's inside the "calculate_refund(inv, recalculation = false)" method.

`self.update_attributes(prev_account_sd_bal: self.account_security_deposit.balance)`

**The proposed solution** is that we update the previous SD balance only at the time of invoice creation.

**Risk:** I've checked it is not being used anywhere else, but @jredfield if you can shed some more light that if it is going to impact any other aspect of the SD then it will great.
 
Thanks

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Feb 19, 2020, 05:59

@kdjstudios Once we confirm that there will be no negative impacts with their proposed solution and tell them to proceed with the estimate, they will provide one. They believe the code changes will not take long, but that testing will have to be extensive.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Feb 19, 2020, 09:42

@amishra @ranand @kdjstudios I have looked at the code and so far don't see any other impacts to SD logic, but I'd like to open this up to discussion on Thursday.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 20, 2020, 08:49

Please post an estimate and then continue with the development. This is approved.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 21, 2020, 02:53

@kdjstudios @jredfield 

Thank you for your pre-approval.

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|X-Small|1|NA|1|High Impact lots of use cases|5|6|

##### Estimate Notes:
* Risk areas identified: `Security Deposit`

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 25, 2020, 24:18

@kdjstudios @jredfield 

* We have completed the development & testing from our end. 
* We have setup branch on the merge server for your testing please test it.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 25, 2020, 10:11

Testing steps:

1. Have account with balance ($234) and SD balance ($500).
2. Edit account and adjust SD. Use appply SD to balance and remainder to invoice. 
3. Created draft invoice. Looks good so far.
![image](/uploads/1ebb43a0ee68a2a12e7630059af7ca77/image.png)
4. Add additional fee ($60) to draft invoice. **[FAILED]**
  - Expects the SD fees to recalculate accordingly on draft invoice.  Deposit Descrease should be (-294.44) and Deposited adjust should be (-205.56).
![image](/uploads/621148a802b08c6b45ba8e94912332e3/image.png)
5. Submit Invoice to see if recalculation is occuring at this step. **[FAILED]**
![image](/uploads/f331324eb6cd28126bb30f9599a4ddf4/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Feb 27, 2020, 24:30

@kdjstudios @jredfield 

I looked into it & found that this scenario was discussed in ticket #28 but I did not find any resolution for this scenario there. This was discussed in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/28#note_35562, http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/28#note_35562 & http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/28#note_35776.

I would suggest that we should create a separate ticket for this issue.


NOTE: In this ticket, we fixed the issue of zero out the SD line items & SD balance on adding & deleting the additional fee in the draft invoice.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 27, 2020, 09:05

@ranand 

This is the same issue, we have so far "half" way fixed it thought. we have prevented it from completely zeroing out, but we still need it to calculate correctly. There is no need for another ticket to fix the same overall issue of the SD fees not working (re-calculating correctly) when adding/removing additional fees on the draft invoice screen.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 3, 2020, 10:43

@ranand @smasih @amishra 

May we have an update please?

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Mar 8, 2020, 23:33

@kdjstudios @jredfield 

Thank you for the update.

* According to your testing & results while adding the Additional Fee the SD fee should decrease/increase accordingly.

* It would be great if you can provide all the test cases for draft invoice calculation. 

* We need to know all the possible test cases & it's expected your results before starting the development.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 9, 2020, 09:59

According to your testing & results while adding the Additional Fee the SD fee should decrease/increase accordingly.
> This is how it was designed in #28, so the test cases might be there as well. 

We need to know all the possible test cases & it's expected your results before starting the development.
> @jredfield Please work with @amishra and create the test cases.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Mar 20, 2020, 04:31

@jredfield @kdjstudios 

Any update on this?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 20, 2020, 08:29

@jredfield 

Please follow up with @ranand and @amishra with creating the test cases.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Mar 27, 2020, 05:55

@kdjstudios Hi Kyle, can you please do a meeting with me on Monday so we can discuss this?  Thx!

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Mar 30, 2020, 07:20

@ranand Pls get the estimate on the firs for us and we'll discuss test cases.  Thx

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Mar 30, 2020, 10:28

@ranand Hi Rashmi,  Can you please explain why we cannot simply call the draft invoice calculation function any time a new line item is added to the draft invoice?  Wouldn't that ensure we always have the correct handling of the SD amount?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 15, 2020, 09:03

Marking this for Thursday discussion, we have not received a response now in over 2 weeks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 16, 2020, 09:29

This needs to be researched and needs to follow the functionality as described by Tim where there is only one function to calculate. This will insure that anytime an update or change happens on the draft invoice this function can be called and everything will recalculate accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 17, 2020, 06:32

@kdjstudios @tim.traylor 

Thank you for the update.

We looked into this and found that the SD calculation depends on the account balance. They never check what fee line items present on the draft invoice.

We have made a short video of this functionality. Please have a look.

https://www.loom.com/share/ea7656a9b0fa45baa2832c6c8396fdc8

Please let us know if we missing something.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 17, 2020, 13:05

@amishra 

Then this is working incorrectly. Please provide and estimate to fix this to work as described by @tim.traylor

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 21, 2020, 05:50

@kdjstudios @tim.traylor 

Thank you for the update.

We have created one test cases as per the current functionality of the SD refund options.

There is 5 option for the SD refund.

1. Take no action. Security Deposit will remain on the account in SA Billing.
1. Keep full Security Deposit. Security Deposit amount will transfer into revenue. This will not show on the next invoice.
1. Apply Security Deposit to Account Balance. Keep remainder of Security Deposit.
1. Apply Security Deposit to Account Balance. Adjust remainder to next invoice.
1. Adjust full Security Deposit to next invoice.


* We have created 3 point test case on the [spreadsheet](https://docs.google.com/spreadsheets/d/136cTbTgOhSdRoilK_j-6qbm6F2MQ9T-w2lWbeOiSZa4/edit?usp=sharing).

* Rest of all the option please create test cases.
 
* Please have a look and fill the expected functionality column.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 21, 2020, 12:24

@amishra @tim.traylor @Robleski 

I am confused with the test cases? We are not changing how the 5 options calculate, we are only changing how SAB implements those calculations. Basically from what the Aavaz team has informed us that we are not currently recalculation when someone makes a change on the draft invoice, but needs to recalculate those when a user does make a change on the draft invoice.

Please discuss this with Tim and Greg in your morning meetings if your team is not understanding this.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 23, 2020, 04:45

@kdjstudios @tim.traylor @Robleski 

* We need to change the calculations to calculate the refund line items as @amishra mentioned in his above comment that the SD refund fee is dependent on the account balance and the credit fees only. 

* I found that in the main SD ticket #28 **Piyush also stated** `"This scenario was never discussed/considered. We are only considering the credit fees for pre-calculating the amount on the pending items. We can have a call with Tim regarding this but according to us, the fees should never be considered if it is draft until it finalized."` in his [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/28#note_35562) under the **Pending Items (Refund Deposit): 35, 36, 42, 43, 60, 61** header.

* After this comment, I don't see any solution to this, or I didn't find any test cases for this scenario. **The last comment on this scenario was of @kdjstudios stating** `"This will require additional follow up by operations to correct this. Please document accordingly."` in his [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/28#note_35776)

* **As SD is a very complex system, before any actual development we just want to make sure that we all are on the same page. So, for a better understanding, we need to have all the test cases beforehand.**

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 23, 2020, 14:11

@ranand @Robleski @tim.traylor 

I agree completely. However since we missed discussing this today, let look at having this discussed during your morning meetings so we do not have to wait another week.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Apr 24, 2020, 07:08

avaaz needs test cases before proceeding; please add to spreadsheet above.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 4, 2020, 10:28

@Robleski @tim.traylor @ranand @amishra 

I am just following up to see what are the updates from your discussion of this in your morning meetings?

**Comment by kdjstudios** [comment]
In GitLab by @ranand on May 11, 2020, 05:16

@kdjstudios @Robleski 

As per our last discussion, we are waiting on the test-cases.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on May 11, 2020, 07:12

greg to follow up with kyle

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on May 11, 2020, 08:52

kyle to post his thoughts

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 11, 2020, 09:03

@ranand @rranjan @amishra @Robleski 

Hey Team may someone please clarify what test cases your team needs? I believe all the 5 type of calculations are to be handled/tested the same that they previously were calculating, so there should be no new test cases needed.

The only update this ticket is doing it to handle recalculating these fees after any change is made on the draft invoice screen. If this is the test case, these should be exactly the same as the ones above, as mentioned by @tim.traylor in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1633#note_58027 we are not changing the calculations, they are simply recalculating the same formula with the updated values of the fees.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on May 12, 2020, 24:55

@kdjstudios @Robleski 

> I believe all the 5 type of calculations are to be handled/tested the same that they previously were calculating, so there should be no new test cases needed.

That's what we have mentioned in our previous comments http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1633#note_58244, http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1633#note_58132, that **we need to change the calculation of SD,** as it is not taking draft invoice line items into consideration other than credit line items.

So, just calling the same method on any change, will not change the SD value.

As, I've mentioned in my previous [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1633#note_58244) that this was discussed in #28 but no resolution or related test-cases are there on #28.

> Hey Team may someone please clarify what test cases your team needs?

We need test cases, where it'll show what and how all line items need to be added in the calculations. Some sample test-cases were given by the @amishra in his above comment.
[link to sample test cases](https://docs.google.com/spreadsheets/d/136cTbTgOhSdRoilK_j-6qbm6F2MQ9T-w2lWbeOiSZa4/edit?usp=sharing).


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 12, 2020, 11:50

@ranand @Robleski 

I am confused with what else is needed from us. Your statement here appears to highlight the only test case needed: "not taking draft invoice line items into consideration other than credit line items."  Simply update this to account for "All" instead of "Credits only". Does that make sense or may you please explain further what is needed from our team?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 14, 2020, 09:45

This was reviewed today and we are setting up a meeting tomorrow to dive further in depth on this.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on May 15, 2020, 09:21

rasdhmi to publish new/add'l test cases. kyle and i to review. then dev work will start.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on May 20, 2020, 06:16

@kdjstudios @Robleski 

As discussed, we have updated the [TestCases sheet](https://docs.google.com/spreadsheets/d/136cTbTgOhSdRoilK_j-6qbm6F2MQ9T-w2lWbeOiSZa4/edit?usp=sharing). Please review them.

* All test cases are written assuming the refund is initiated from the invoice.

In addition to this, I have a query:
* Let's say, we initiated the SD refund from the account edit page.
* Created an invoice, on which this SD refund will get applied.
* At the time of invoice creation, will we recalculate the refund to accommodate the invoice balance or not?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 20, 2020, 09:28

@ranand @Robleski 

Any updates or changes to an invoice will recalculate the SD line item amounts.

After reviewing the Test cases it looks like you only created these for one of the refund/adjust options, there are five total I believe. Do you need to create test cases for each option, or are you able to use the existing test cases for those refund/adjust options?

**Comment by kdjstudios** [comment]
In GitLab by @ranand on May 21, 2020, 24:08

@kdjstudios @Robleski 

> After reviewing the Test cases it looks like you only created these for one of the refund/adjust options, there are five total I believe. Do you need to create test cases for each option, or are you able to use the existing test cases for those refund/adjust options?

* We've fixed the [TestCases sheet](https://docs.google.com/spreadsheets/d/136cTbTgOhSdRoilK_j-6qbm6F2MQ9T-w2lWbeOiSZa4/edit?usp=sharing). 
* It has test-cases with the 4 refund options.
* For "Take no action. Security Deposit will remain on the account in SA Billing.", we didn't add any test-cases as no action is performed for it.

> Any updates or changes to an invoice will recalculate the SD line item amounts.

So, SD fees which originated from the account edit page, at the time invoice creation, it will get recalculated.


Please, review the test-cases.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 21, 2020, 13:10

@ranand @Robleski 

Look good to me. Please provide an estimate to accomplish this.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on May 26, 2020, 05:12

@kdjstudios @Robleski 

Thank you for the update.

Please find the below estimate:

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-1|4|Aware of Concepts, implementation tricky|7|High Impact lots of use cases|9.0|16|

##### Estimate Notes:
* Risk areas identified: `Security Deposit`
* Work plan: `Change the existing Security Deposit functionality.`
* Test cases: `Mentioned in` [here](https://docs.google.com/spreadsheets/d/136cTbTgOhSdRoilK_j-6qbm6F2MQ9T-w2lWbeOiSZa4/edit#gid=0)

`Note: This is the rough estimate it might take more or less time.`

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 26, 2020, 15:21

@Robleski 

Please let us know if this is approved to have the team start working on.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on May 26, 2020, 18:35

this is approved for avaaz to start work

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Jun 1, 2020, 07:15

Failed AN Acceptance testing. Back in Avaaz dev/dev. Expecting to be ready for AN acceptance testing on 6/2.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 4, 2020, 06:26

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 4, 2020, 06:28

@kdjstudios @Robleski 

We have completed the development and set up a branch on the merge server for your testing.


* Test case is mentioned on the top under the description section.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 5, 2020, 08:02

@kdjstudios @Robleski 

Please test it and give your valuable feedback.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 5, 2020, 09:20

@amishra 

Thanks for the updates I was able to test this and edit the draft invoice a few items. The SD fees adjust accordingly. Marking for ready, if we can release this Saturday wonderful, if not we will need to wait until the next release.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 6, 2020, 05:55

@kdjstudios @Robleski 

Thank you for the update.

We have successfully released on both of the production servers.


thanks

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Jun 9, 2020, 16:46

Dev completed. Expecting to be ready for AN acceptance testing by 6/4.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Jun 9, 2020, 16:47

In AN acceptance testing

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 10, 2020, 08:35

@Robleski 

This has already been developed, tested, and released to production. You might have posted that on the wrong ticket.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 10, 2020, 14:25

I wrote:

> We just wanted to follow up and let you know that this last Saturday (6/6/20) the SA Billing support team was able to correct the bug in the system where the SD would disappear. Please let us know if you have any questions or concerns.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 23, 2020, 15:08

closed
